### ISSUE TITLE
Prepare Lawnchair 11 for F-Droid support

### ISSUE THREAD

**Issue by divadsn** [Deletescape-Media/Lawnchair#1473]
Before someone else opens it.

Continuing the discussion from #240 and #1157.

**Comment by divadsn** [comment]
Leaving this here as a note:
> paphonb, [09.04.19 11:28]
  The thing is that quickstep thing requires an open source prebuilt
  That we can’t inline because it literally takes hours to build

Will probably contact someone from F-Droid to ask if they are willing to do that.

**Comment by licaon-kter** [comment]
What is a "an open source prebuilt" ?

Also, no Tags?

**Comment by nyancrimew** [comment]
@licaon-kter The framework.jar and SystemUIShared.jar in the repository are built from AOSP sources (the modifications can be found in our organization) a build of the framework jar takes multiple hours and the SystemUI Shared library also at least 30 minutes, thus they are included as jars here

Tags will be created again for stable releases and possibly also for betas, though we'll have to first discuss this internally

**Comment by divadsn** [comment]
@licaon-kter Tags are for Play Store releases planned.

**Comment by licaon-kter** [comment]
Ok, right, so no F-Droid build anytime soon, if ever :(

**Comment by divadsn** [comment]
Why that? If F-Droid allows to use a prebuilt library like any other dependency?

**Comment by licaon-kter** [comment]
@divadsn All the other deps are either straight build from source at build time, or taken from a maven repo where only open-source artefacts are published.

What case is this?

**Comment by nyancrimew** [comment]
@licaon-kter We could obviously also push these Jars to a Maven repo if that somehow changes anything about the trust in them

**Comment by licaon-kter** [comment]
@deletescape Yes, that would be fine.

It needs to be one of these allowed maven repos: https://gitlab.com/fdroid/fdroidserver/blob/master/fdroidserver/scanner.py#L99

**Comment by nyancrimew** [comment]
@licaon-kter Alright, we'll look into it on the weekend

**Comment by licaon-kter** [comment]
@divadsn Tags allow F-Droid to autoupdate on each release.

**Comment by thw0rted** [comment]
This is probably the wrong place to ask but when should we expect to see a v2 build on the Play Store?  I was going to ask on #1462 but that's been locked.  (If this is documented somewhere, or if there's a more appropriate forum to ask questions like this, I'd appreciate being pointed in the right direction.)

**Comment by divadsn** [comment]
@thw0rted Use the search, we already gave a approximate ETA, probably this weekend.

**Comment by thw0rted** [comment]
Thanks for the prompt response @divadsn!

I searched for "v2", "google play", "play store", "schedule", and haven't found any issues whose titles suggest they'd be a good place to find information about the Play Store release schedule.  The README does not address the build/release process or schedule, nor does the FAQ on the wiki.  I'm not complaining!  I just wanted to try to answer my own question without bugging the devs about it.

**Comment by TacoTheDank** [comment]
Perhaps an exception can be made for Lawnchair? Dunno what else to do in this scenario...

**Comment by thw0rted** [comment]
Taco, it looks like licoan posted a solution already.  The custom built JARs need to be published as a first class artifact in a trusted Maven repo.  Basically, I think they need to be their own "project", then Lawnchair can depend on `lawnchair-framework` or whatever they decide to call it.

**Comment by licaon-kter** [comment]
@TacoTheDank Eh... all devs would then want exceptions...

**Comment by TacoTheDank** [comment]
@licaon-kter Mmm, that could happen...

@thw0rted Well that's good.

**Comment by nikolowry** [comment]
Hosting your own F-Droid repo is an alternative if certain build deps are preventing a release in the official repo -- microG, The Guardian and bromite have all taken this route

**Comment by PopeRigby** [comment]
Why is the upstream not entirely free?

**Comment by ClockGen** [comment]
@deletescape @divadsn Hello, any progress on releasing jars to maven repos?

**Comment by divadsn** [comment]
They should be already prepared to be uploaded, @deletescape is uploading
them.

**Comment by RichyHBM** [comment]
Has there been any progress with this?

**Comment by losuler** [comment]
Someone recently created a fork called [Librechair](https://gitlab.com/oldosfan/librechair) which may be helpful.

**Comment by suphon-t** [comment]
@losuler The main problem (the framework jar) still can’t be solved yet.

**Comment by RichyHBM** [comment]
Sounds like that would also have the same issues getting onto FDroid, plus I'd quite like the sesame integration!

**Comment by ghost** [comment]
@paphonb you can go the barbarian way of uploading the platform jars to MvnCentral

**Comment by rugk** [comment]
Any news?

**Comment by suphon-t** [comment]
> Any news?

There's been no progress on this as it is not a priority

**Comment by rugk** [comment]
Well… [based on the upvotes of the issues in this repo](https://github.com/LawnchairLauncher/Lawnchair/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) this actually should be top priority 1…
It seems this is what most of your users actually want. :thinking: 
Given that, I hope this could influence your priorities a little. :smiley:

**Comment by Roy-Orbison** [comment]
> Alright, we'll look into it on the weekend

@deletescape Which one?

**Comment by IzzySoft** [comment]
@deletescape @divadsn if you could attach the (signed release-) APK to `releases/` here, I could take it into [my repo](https://apt.izzysoft.de/fdroid/) until a solution for the official repo is found. What do you think?

**Comment by ClockGen** [comment]
@divadsn @deletescape Please publish the jars in one of f-droid supported maven repos https://gitlab.com/fdroid/fdroidserver/blob/master/fdroidserver/scanner.py#L99

UPD: Unfortunately the librechair fork seems to be not be maintained anymore, and it still didn't achieve its goal (to be fully libre and f-droid compatible).

**Comment by jrtberlin** [comment]
Has there been any progress since November?

**Comment by ildar** [comment]
Since November? which year?

**Comment by rradar** [comment]
unstale this :robot:

**Comment by RanceJustice** [comment]
Adding my support and interest in getting the latest builds of Lawnchair onto F-Droid!    I'm to understand that Lawnchair is one of the most modern and powerful  libre/open source  launchers for Android at current, offering Pixel Launcher on Android 10 features and then going beyond.    On top of it, it compares favorably to proprietary ones.   While its good to see a launcher like that on the Play Store, it would be nice to see it in FDroid for both those who prioritize FLOSS solutions  when possible and for newcomers wondering  what libre/open source apps can serve their needs. 

So at the moment, what are the main blockers preventing official FDroid release?   Is it still the .jar issue?  If so, is the idea of putting it into  a libre FDroid supported Maven repo still  an acceptable  / the best solution for allowing Lawnchair to be carried by the official FDroid repo?  

Alternatively,  would it be helpful to offer Lawnchair's latest and full featured builds via another FDroid compatible 3rd party repo?  As others have mentioned, the Guardian Project , MicroG, Bitwarden and others utilize these, and it could be either a new one specifically hosted for Lawnchair or if you find another one that is bother technically and ideologically compatible, perhaps that would also work.  A 3rd party repo could  be a temporary asset while inclusion in main FDroid is being sorted, or if necessary a compromise if the aforementioned doesn't work out for one reason or another. 

Thank you.

**Comment by issu-shooter** [comment]
The Librechair fork was migrated from gitlab to sourcehut, it's a rather unpopular platform. According to a issue at gitlab repo the author left gitlab because it has trackers.

**Comment by ildar** [comment]
I've noticed a sign of activity, which is very good :-) (
https://github.com/LawnchairLauncher/lawnchair/releases/tag/v11.0-alpha.1)
Yet is seems to lose FOSS compatibility, see details :
https://gitlab.com/fdroid/rfp/-/issues/1025#note_540337923
@patrykmichalik , can you fix this? thanks in advance !

**Comment by msfjarvis** [comment]
> Yet is seems to lose FOSS compatibility, see details : [gitlab.com/fdroid/rfp/-/issues/1025#note_540337923](https://gitlab.com/fdroid/rfp/-/issues/1025#note_540337923) @patrykmichalik , can you fix this? thanks in advance !

The GMS dependency can potentially be dropped, but the prebuilt JARs are included in that form because it is impossible to build them alongside the app.

**Comment by ildar** [comment]
On Thu, Apr 1, 2021 at 2:49 PM Harsh Shandilya ***@***.***>
wrote:

> Yet is seems to lose FOSS compatibility, see details :
> gitlab.com/fdroid/rfp/-/issues/1025#note_540337923 @patrykmichalik
> <https://github.com/patrykmichalik> , can you fix this? thanks in advance
> !
>
> The GMS dependency can potentially be dropped
>
That's great! possible to make (or restore) the foss / fdroid flavor ?

> but the prebuilt JARs are included in that form because it is impossible
> to build them alongside the app.
>
 Are all of the jars' sources available?

**Comment by msfjarvis** [comment]
>> The GMS dependency can potentially be dropped
>
> That's great! possible to make (or restore) the foss / fdroid flavor ?

It's not a priority but yes it can be done.


>> but the prebuilt JARs are included in that form because it is impossible to build them alongside the app.
>
> Are all of the jars' sources available?

They are, F-Droid does not care for that. They want dependencies to only originate from a set of repositories they trust and enforce the use of.

**Comment by ildar** [comment]
On Thu, Apr 1, 2021 at 3:28 PM Harsh Shandilya wrote:

> The GMS dependency can potentially be dropped
>
> That's great! possible to make (or restore) the foss / fdroid flavor ?
>
> It's not a priority but yes it can be done.
>
That would be mandatory (required by F-Droid maintainers)

> but the prebuilt JARs are included in that form because it is impossible
> to build them alongside the app.
>
> Are all of the jars' sources available?
>
> They are, F-Droid does not care for that. They want dependencies to only
> originate from a set of repositories they trust and enforce the use of.
>
Not exactly. Some jars / components for the target project can be built
with the project itself. So this part is solvable allright.
We just need all the source URLs.

Should I open a separate FR for foss flavor?

**Comment by msfjarvis** [comment]
> On Thu, Apr 1, 2021 at 3:28 PM Harsh Shandilya wrote: The GMS dependency can potentially be dropped That's great! possible to make (or restore) the foss / fdroid flavor ? It's not a priority but yes it can be done.
> That would be mandatory (required by F-Droid maintainers)

Yes, mandatory for them, and not a priority for us. We have a laundry list of features to work on before we start prioritising distribution on F-Droid.

> but the prebuilt JARs are included in that form because it is impossible to build them alongside the app. Are all of the jars' sources available? They are, F-Droid does not care for that. They want dependencies to only originate from a set of repositories they trust and enforce the use of.
> Not exactly. Some jars / components for the target project can be built with the project itself. So this part is solvable allright. We just need all the source URLs. Should I open a separate FR for foss flavor?

The source for the `sysui_shared.jar` and `sysui_statslog.jar` dependencies is available [here](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/android11-release/packages/SystemUI/shared/), `plugin_core.jar` is built from [here](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/android11-release/packages/SystemUI/plugin/), and `framework.jar` is built from [here](https://android.googlesource.com/platform/frameworks/base/+/refs/heads/android11-release/core/). I would love for you to open a PR that builds these inline.

**Comment by divadsn** [comment]
As I wrote on GitLab, we'll see how it will go. As for development side of things, we will this time try to implement new features with F-Droid build flavor in mind, so in case we add some proprietary feature, we will have an alternative for that.

**Comment by ClockGen** [comment]
@divadsn as mentioned before, the dirty workaround of uploading these (open source) jars to any of the supported maven repos (https://gitlab.com/fdroid/fdroidserver/-/blob/master/fdroidserver/scanner.py#L152) will allow us to use them.

**Comment by Shayan-To** [comment]
Any estimate on when Lawnchair will be available on F-Droid?

**Comment by julianfairfax** [comment]
Given that Lawnchair 1 was on F-Droid, and you've mentioned implementing a prebuilt copy of the Android system framework from AOSP source as being a solution, isn't it time Lawnchair 2 gets added to F-Droid? This is, after all, an (/the best) open source launcher and one of the only applications on my phone that still isn't in F-Droid, which is disappointing to say the least. Thank you for all the work you've done on this application and hopefully I'll see this application on F-Droid soon! (For anyone still wanting to download it, you can get it from https://www.apkmirror.com/apk/deletescape/lawnchair-launcher)

**Comment by licaon-kter** [comment]
@julianfairfax why use some random site instead of https://github.com/LawnchairLauncher/lawnchair/releases ?

**Comment by julianfairfax** [comment]
> @julianfairfax why use some random site instead of https://github.com/LawnchairLauncher/lawnchair/releases ?

Well I'm specifically looking to download Lawnchair 2, which is the latest version in the Google Play Store (and Lawnchair 1 is the latest version in F-Droid). Lawnchair 11 (is that what they're calling it now) also don't work on my device since it requires Android 11 and is also a pre-release (nothing wrong with that, I just want a stable version). And I wouldn't consider it a random site a random site since https://lawnchair.app/ links to another page on that site as an official download method (although the link on their website brings me to the pre-release versions, not the stable versions).

**Comment by msfjarvis** [comment]
> Given that Lawnchair 1 was on F-Droid, and you've mentioned implementing a prebuilt copy of the Android system framework from AOSP source as being a solution, isn't it time Lawnchair 2 gets added to F-Droid? This is, after all, an (/the best) open source launcher and one of the only applications on my phone that still isn't in F-Droid, which is disappointing to say the least. Thank you for all the work you've done on this application and hopefully I'll see this application on F-Droid soon! (For anyone still wanting to download it, you can get it from [apkmirror.com/apk/deletescape/lawnchair-launcher](https://www.apkmirror.com/apk/deletescape/lawnchair-launcher))

We use prebuilts because we can't be building all of AOSP within the Lawnchair build system. F-Droid is against prebuilt binaries, which is why this issue is still open.


As a general note, repeatedly going "ETA pls" does not have the effect you might think it does, so cutting down on that spam would be appreciated.
